### PR TITLE
Apply rope on k earlier for efficiency

### PR DESCRIPTION
**Pull request by jackzhxng**
#### Context
By performing rotational embeddings early, we can squeeze out a bit more performance. Comparing finetuning statistics on Llama3.1-Instruct 8B w LoRA on GPU (full stats are in the "Appendix" section at the end of this pr description):
- Tokens per second increased from `1246.51852/s` ->`1316.82931/s` (**+70 tokens per second**)
- Loss stayed about the same (`1.48158` -> `1.48607`)
- Peak active memory stayed the same (`18.14733` -> `18.14733`)

#### Changelog
Does rope on an unexpanded k that is still `[b, s_y, n_kv, h_d]` instead of `[b, s_y, n_h, h_d]`.

#### Test plan
- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
No public API changes

#### Appendix
Stats from one epoch of finetuning Llama3.1-Instruct 8B **prior** to this PR:
```
1|25|Loss: 1.481581211090088: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [12:49<00:00, 30.78s/it]
wandb:                                                                                
wandb: 
wandb: Run history:
wandb:               global_step ▁▁▂▂▂▂▃▃▃▄▄▄▅▅▅▅▆▆▆▇▇▇▇██
wandb:                      loss ▆▆▇▇▆▆▇▇▅▇▅▅█▅▆▆▅▅▄▃▃▃▂▂▁
wandb:                        lr ▁▁▂▂▂▂▃▃▃▄▄▄▄▅▅▅▆▆▆▇▇▇▇██
wandb:        peak_memory_active ▅▆▇█▆█▇▆▄██▆▅▆▁▆▆▇█▄▇▇▃▅▇
wandb:         peak_memory_alloc ▅▆▇█▆█▇▆▄██▆▅▆▁▆▆▇█▄▇▇▃▅▇
wandb:      peak_memory_reserved ▁▁▃██████████████████████
wandb: tokens_per_second_per_gpu ▄▅▅▅▅▃▂▁▆▃▅█▃▆▁▄▂▅▂▃▄▅▃▂▅
wandb: 
wandb: Run summary:
wandb:               global_step 25
wandb:                      loss 1.48158
wandb:                        lr 7e-05
wandb:        peak_memory_active 18.14733
wandb:         peak_memory_alloc 18.14733
wandb:      peak_memory_reserved 19.43555
wandb: tokens_per_second_per_gpu 1246.51852
wandb: 
wandb: 🚀 View run lilac-puddle-14 at: https://wandb.ai/dvorjackz-meta/torchtune/runs/21e7roul
wandb: ⭐️ View project at: https://wandb.ai/dvorjackz-meta/torchtune
wandb: Synced 5 W&B file(s), 0 media file(s), 2 artifact file(s) and 1 other file(s)
wandb: Find logs at: /tmp/wandb/run-20240916_071220-21e7roul/logs
```

Stats from one epoch of finetuning Llama3.1-Instruct 8B **after** this PR:
```
1|25|Loss: 1.48606538772583: 100%|██████████████████████████████| 25/25 [12:34<00:00, 30.17s/it]
wandb:                                                                                
wandb: 
wandb: Run history:
wandb:               global_step ▁▁▂▂▂▂▃▃▃▄▄▄▅▅▅▅▆▆▆▇▇▇▇██
wandb:                      loss ▆▆▇▆▆▆▇▇▅▇▅▅█▆▆▆▅▅▄▃▃▂▂▂▁
wandb:                        lr ▁▁▂▂▂▂▃▃▃▄▄▄▄▅▅▅▆▆▆▇▇▇▇██
wandb:        peak_memory_active ▅▆▇█▆█▇▆▄██▆▅▆▁▆▆▇█▄▇▇▃▅▇
wandb:         peak_memory_alloc ▅▆▇█▆█▇▆▄██▆▅▆▁▆▆▇█▄▇▇▃▅▇
wandb:      peak_memory_reserved ▁▁▃██████████████████████
wandb: tokens_per_second_per_gpu ▅▇▄▆▆▅▅▂▆▄▇█▂▆▂▄▁▄▃▃▄▅▄▄▆
wandb: 
wandb: Run summary:
wandb:               global_step 25
wandb:                      loss 1.48607
wandb:                        lr 7e-05
wandb:        peak_memory_active 18.14733
wandb:         peak_memory_alloc 18.14733
wandb:      peak_memory_reserved 19.43555
wandb: tokens_per_second_per_gpu 1316.82931
wandb: 
wandb: 🚀 View run azure-forest-13 at: https://wandb.ai/dvorjackz-meta/torchtune/runs/uw4eja2j
wandb: ⭐️ View project at: https://wandb.ai/dvorjackz-meta/torchtune
wandb: Synced 4 W&B file(s), 0 media file(s), 3 artifact file(s) and 1 other file(s)
wandb: Find logs at: /tmp/wandb/run-20240916_064527-uw4eja2j/logs
```

### CODE REVIEW

File: torchtune/modules/attention.py [hunk 1]
```diff
@@ -259,27 +259,30 @@ def forward(
             k = self.k_proj(y)
             v = self.v_proj(y)
 
+            # Apply positional embeddings
+            # k: [b, s_y, n_kv, h_d]
+            k = k.view(b, s_y, self.num_kv_heads, self.head_dim)
```
Comment (kimishpatel) [inline torchtune/modules/attention.py:264]:
```suggestion
            k = k.view(b, s_y, -1, self.head_dim)
```

Dont change the -1 part?

Comment (dvorjackz) [reply torchtune/modules/attention.py:264]: Just curious what's the reason?

**Comment by felipemello1** [comment]
Thanks for the PR! Nice catch! Do you think you can run it with vs without your changes and show the improvement + no change in loss? It will make approving the PR easy to approve

If you use weights and biases, its probably the easiest way to take a screenshot (you can login with your gmail, get the token, pip install wandb, wandb login, insert your token)

then you can run your config like this:
```tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=25  epochs=1 metric_logger=torchtune.training.metric_logging.WandBLogger log_peak_memory_stats=True```

File: torchtune/modules/attention.py (hunk 1 above)
Comment (kimishpatel) [reply torchtune/modules/attention.py:264]: @dvorjackz are you asking me why i suggested the change? If, that was just to preserve what was there before.

But if you are asking why this diff, then maybe summary explains

**Review by kimishpatel — APPROVED** [review]
looks good to me and i will stamp but hopefully other folks in tune are ok with this

File: torchtune/modules/attention.py [hunk 2]
```diff
@@ -259,27 +259,30 @@ def forward(
...
+            # View + expand + reshape bring num_kv_heads to num_heads for k and v
+            # to match q.
+
             # k: [b, s_y, n_kv, 1, h_d]
             # v: [b, s_y, n_kv, 1, h_d]
             k = k.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
             v = v.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
 
-            # if needed, expand the key and value tensors to have the same shape
+            # Expand the key and value tensors to have the same shape
```
Comment (ebsmothers) [inline torchtune/modules/attention.py:276]: I think the "if needed" is still relevant here. This class is also used for non-Llama models that will not hit this path (e.g. [Phi3](https://github.com/pytorch/torchtune/blob/ee343e61804f9942b2bd48243552bf17b5d0d553/torchtune/models/phi3/_model_builders.py#L35-L36))

**Comment by ebsmothers** [comment]
Yeah +1 to @felipemello1's suggestion. Otherwise this looks good to me, can stamp once you update the summary with the results

**Comment by kimishpatel** [comment]
@ebsmothers i tried to find your gh handle for tagging. now i know

**Comment by felipemello1** [comment]
cool! thanks! Lets merge it after tests are done

**Comment by dvorjackz** [comment]
@felipemello1 need another approval since there have been changes since Kimish's approval 👀

**Review by felipemello1 — APPROVED** [review]
